### PR TITLE
Fixes the problem of being unable to connect to HiveServer2 4.0.1 with Zookeeper Service Discovery enabled

### LINKED ISSUE #22777: Hive Connection Error with ZooKeeper
### Description

Hi, 

I created new connection to hive and connecting with zookeeper quorum, it is throwing below error. When I create direct connection to hive server, it is working fine. Can you please help fixing this error?

Connection URL (Working Fine):

jdbc:hive2://host:10000/db_name

Connection URL (Throwing Error):

jdbc:hive2://hostname:2181/;serviceDiscoveryMode=zooKeeper;zooKeeperNamespace=hiveserver2

Unexpected driver error occurred while connecting to the database
'com.google.common.util.concurrent.ListeningExecutorService com.google.common.util.concurrent.MoreExecutors.sameThreadExecutor()'
'com.google.common.util.concurrent.ListeningExecutorService com.google.common.util.concurrent.MoreExecutors.sameThreadExecutor()'

### DBeaver Version

Enterprise Version 23.3.0.202312121701

### Operating System

Windows 11

### Database and driver

Database Name: Apache Hive
Driver: hive-jdbc-uber-2.6.5.0-292.jar

### Steps to reproduce

1. Create new connection with Apache Hive
2. give below URL (Connect through ZooKeeper)
jdbc:hive2://hostname:2181/;serviceDiscoveryMode=zooKeeper;zooKeeperNamespace=hiveserver2
3. Test connection, it will throw the error

### Additional context

When connection created direct to Hive, it would work fine.
jdbc:hive2://hostname:10000/db_name

### PR DESCRIPTION
**Pull request by linghengqian**
- Fixes #22777 .
- Fixes the problem of being unable to connect to HiveServer2 4.0.1 with Zookeeper Service Discovery enabled. For more information, see https://github.com/dbeaver/dbeaver/issues/22777#issuecomment-2264969480 .
- image

### CODE REVIEW

**Comment by E1izabeth** [comment]
Thank you for contribution!

File: plugins/org.jkiss.dbeaver.ext.hive/plugin.xml
Comment (linghengqian) [inline plugins/org.jkiss.dbeaver.ext.hive/plugin.xml]: - To be honest I can't find any explanation in the dbeaver documentation on how to define the `classifier` property in this place. Should this be `maven:/org.apache.hive:hive-jdbc:RELEASE[4.0.0-standalone]`, or `maven:/org.apache.hive:hive-jdbc:standalone:RELEASE[4.0.0]`? The relevant JAR is located at https://repo1.maven.org/maven2/org/apache/hive/hive-jdbc/4.0.0/hive-jdbc-4.0.0-standalone.jar .

**Review by E1izabeth — DISMISSED** [review]

File: plugins/org.jkiss.dbeaver.ext.hive/plugin.xml
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ext.hive/plugin.xml]: No need to add `-standalone` here. Please left just `maven:/org.apache.hive:hive-jdbc:RELEASE[4.0.0]`

Comment (linghengqian) [reply plugins/org.jkiss.dbeaver.ext.hive/plugin.xml]: > No need to add `-standalone` here. Please left just `maven:/org.apache.hive:hive-jdbc:RELEASE[4.0.0]`

- Resolved in the latest commit. But from my perspective, `maven:/org.apache.hive:hive-jdbc:RELEASE[4.0.0]` points to,
```xml
<dependency>
            <groupId>org.apache.hive</groupId>
            <artifactId>hive-jdbc</artifactId>
            <version>4.0.0</version>
</dependency>
```
- But what is actually needed is,
```xml
<dependency>
            <groupId>org.apache.hive</groupId>
            <artifactId>hive-jdbc</artifactId>
            <version>4.0.0</version>
            <classifier>standalone</classifier>
</dependency>
```

**Comment by LonwoLonwo** [comment]
Hello @linghengqian 

We are concerned about vulnerabilities in https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-common/3.3.6

So, the option from your side looks ok for us:

> Once apache/hive:4.0.1 is released, the dependency on org.apache.hadoop:hadoop-common:3.3.6 will be unnecessary. See the relevant PR https://github.com/apache/hive/pull/5313 and the relevant mailing list thread https://lists.apache.org/thread/63lr45kyh78s64spwsjxjy8zdyzprnz .

We will wait for the next release and check the new vulnerability scan report.

**Comment by linghengqian** [comment]
> Hello @linghengqian
> 
> We are concerned about vulnerabilities in https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-common/3.3.6
> 
> So, the option from your side looks ok for us:
> 
> > Once apache/hive:4.0.1 is released, the dependency on org.apache.hadoop:hadoop-common:3.3.6 will be unnecessary. See the relevant PR [apache/hive#5313](https://github.com/apache/hive/pull/5313) and the relevant mailing list thread https://lists.apache.org/thread/63lr45kyh78s64spwsjxjy8zdyzprnz .
> 
> We will wait for the next release and check the new vulnerability scan report.

- That's not a problem. Should I close the current PR first? 
- It may take another 2 months to release apache/hive 4.0.1, because it needs to wait for the release of Tez `0.10.4`. See https://lists.apache.org/thread/63lr45kyh78s64spwsjxjy8zdyzprnz1 . By the way, the Docker Image `apache/hive:4.0.0` of HiveServer2 uses Tez in Local mode by default, which does have an impact.
- Even in the master branch of apache/hive, the Uber JAR of HiveServer2 JDBC Driver can only be used with Zookeeper Client `3.8.x`, and it is not compatible with Zookeeper Client `3.9.2`. This was reported by me as https://issues.apache.org/jira/browse/HIVE-28445 . I am not a committer of apache/hive, so I would say that I don’t know how the hive community plans to solve this problem. Maybe by the time 4.1.0 is released, users will need to manually import the Maven dependency of Zookeeper Client.
```xml
<dependency>
    <groupId>org.apache.zookeeper</groupId>
    <artifactId>zookeeper</artifactId>
    <version>3.9.2</version>
</dependency>
<dependency>
    <groupId>org.apache.curator</groupId>
    <artifactId>curator-framework</artifactId>
    <version>5.7.0</version>
</dependency>
```
- For those interested in collaborating with the Hive community, you can temporarily test the Thin JAR I built for the HiveServer2 JDBC Driver at https://github.com/linghengqian/hive-server2-jdbc-driver . The related artifacts are being used in nativeTest under the GraalVM Native Image of apache/shardingsphere. This test artifact is compatible with Zookeeper Client 3.9.2.

**Review by linghengqian — COMMENTED** [review]
- @E1izabeth 
- I have updated the current PR to HiveServer2 JDBC Driver 4.0.1 and can connect to the local instance using `jdbc:hive2://localhost:2181/;serviceDiscoveryMode=zooKeeper;zooKeeperNamespace=hiveserver2;` fine in the snapcraft package of the older version of dbeaver-ce.
- image
- But I still don't understand the grammatical rules here.
> `@E1izabeth`: No need to add -standalone here. Please left just maven:/org.apache.hive:hive-jdbc:RELEASE[4.0.0]

- I don't think the following maven POM definition is equivalent to `maven:/org.apache.hive:hive-jdbc:RELEASE[4.0.1]`.

```xml
        <dependency>
            <groupId>org.apache.hive</groupId>
            <artifactId>hive-jdbc</artifactId>
            <version>4.0.1</version>
            <classifier>standalone</classifier>
        </dependency>
```

**Comment by ivgag** [comment]
@linghengqian
> I don't think the following maven POM definition is equivalent to maven:/org.apache.hive:hive-jdbc:RELEASE[4.0.1].

You are right. The classifier can be defined like this: `maven:/org.apache.hive:hive-jdbc:standalone:RELEASE[4.0.1]`

**Review by linghengqian — COMMENTED** [review]
- All issues have been resolved.

**Review by ivgag — APPROVED** [review]

**Comment by arhayka** [comment]
@linghengqian Thank you for resolving all the issues!
During testing, we found that this Hive version is incompatible with previous versions.
Before merging the PR, we’ll need to create an additional connector for this version of Hive. We plan to address this in the next sprint (24.2.5) within the next two weeks.

**Comment by linghengqian** [comment]
> @linghengqian Thank you for resolving all the issues!
> During testing, we found that this Hive version is incompatible with previous versions.
> Before merging the PR, we’ll need to create an additional connector for this version of Hive. We plan to address this in the next sprint (24.2.5) within the next two weeks.

- That's not a problem. But I think the new connector module's PR should not need me to handle, right?👀 I assume that the new connector module is actually only for the EOL Hive `2.x` or Hive `3.x`, and my personal concern for the unit test on the Apache Shardingsphere side is only Hive `4.0.1`.
- For the HiveServer2 JDBC driver `4.0.1`, the use of Hadoop `3.3.6` with the CVE is still present. Support for Hadoop `3.4.0` is still being discussed at https://github.com/apache/hive/pull/5500 , as it involves multiple disruptive changes to Hadoop `3.4.0`.

**Comment by arhayka** [comment]
@linghengqian, you are correct, we’ve already added a new Hive connector, and it’s available in v24.2.5.
So, I’ve closed this PR.

Thank you once again for your contribution and research. We hope to consolidate everything into a single Hive connector in the future, but for now, it will remain as it is.

Screenshot 2024-11-18 at 12 22 50

**Pull request closed without merging**
